### PR TITLE
feat(backend/frontend/shared): add real time notifications tm-260

### PR DESCRIPTION
**Pull request by Sytnikovakatya**
Implement real-time notifications

### CODE REVIEW

File: packages/shared/src/modules/notifications/libs/types/notification-with-receiver-id-response-dto.type.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import { type NotificationResponseDto } from "./notification-response-dto.type.js";
+
+type NotificationWithReceiverIdResponseDto = Record<"receiverId", number> &
+	Record<"notification", NotificationResponseDto>;
```
Comment (fshabanov) [inline packages/shared/src/modules/notifications/libs/types/notification-with-receiver-id-response-dto.type.ts:4]:
```suggestion
type NotificationWithReceiverIdResponseDto = {
  receiverId: number;
	notification: NotificationResponseDto>;
}
```
Why not like this?

**Review by fshabanov — CHANGES_REQUESTED** [review]

**Review by Oleksandra-Masiuk — CHANGES_REQUESTED** [review]

**Review by GEOFARL — APPROVED** [review]

File: apps/backend/src/libs/modules/socket/socket-service.module.ts [hunk 2]
```diff
@@ -17,6 +17,16 @@ class SocketService {
 		});
 	}
 
+	private notificationsHandler(socket: Socket): void {
```
[the remark this answers is not in the document]
Comment (Sytnikovakatya) [reply apps/backend/src/libs/modules/socket/socket-service.module.ts:20]: Yes, sure

**Review by Oleksandra-Masiuk — APPROVED** [review]

File: apps/backend/src/modules/friends/friend.service.ts [hunk 3]
```diff
@@ -52,12 +60,19 @@ class FriendService {
...
 			userId: followerUserId,
 		});
 
+		this.socketService.emitMessage({
+			event: SocketEvent.NOTIFICATIONS_NEW_FOLLOWER,
+			payload: notification,
+			receiversIds: [String(notification.userId), String(receiverId)],
+			targetNamespace: SocketNamespace.NOTIFICATIONS,
+		});
+
```
Comment (MappariChan) [inline apps/backend/src/modules/friends/friend.service.ts:75]: I think you should do this inside of the notification service and inject this socket service into it

File: packages/shared/src/libs/modules/socket/libs/enums/socket-event.enum.ts [hunk 4]
```diff
@@ -3,6 +3,9 @@ const SocketEvent = {
 	CHAT_JOIN_ROOM: "chat-join-room",
 	CHAT_LEAVE_ROOM: "chat-leave-room",
 	CONNECTION: "connection",
+	NOTIFICATIONS_JOIN_ROOM: "notifications-join-room",
+	NOTIFICATIONS_LEAVE_ROOM: "notifications-leave-room",
+	NOTIFICATIONS_NEW_FOLLOWER: "notifications-new-follower",
```
Comment (MappariChan) [inline packages/shared/src/libs/modules/socket/libs/enums/socket-event.enum.ts:8]:
```suggestion
	NEW_NOTIFICATION: "new-notification",
```
I think there will be only one socket message handler for all type of notifications, so I think this enum value would be better

File: apps/frontend/src/libs/modules/store/middlewares/notifications-socket.middleware.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+	return ({ dispatch }) => {
+		notificationsSocketInstance.on(
+			SocketEvent.NOTIFICATIONS_NEW_FOLLOWER,
+			() => {
+				void dispatch(userNotificationsActions.getUserNotifications());
+				void dispatch(
+					userNotificationsActions.checkHasUserUnreadNotifications(),
+				);
+			},
+		);
```
Comment (MappariChan) [inline apps/frontend/src/libs/modules/store/middlewares/notifications-socket.middleware.ts:30]: I think you should retrieve notification from the payload of the socket message and set it to the state of the notification slice, but not sending new request on the data

File: apps/backend/src/modules/friends/friend.service.ts (hunk 3 above)
Comment (Sytnikovakatya) [reply apps/backend/src/modules/friends/friend.service.ts]: Ok, I will move it to the notification service

**Pull request merged**